File: core/ui/api/src/androidMain/kotlin/com/isharaw/kmpproj/core/FeatureGate.kt
package com.isharaw.kmpproj.core

import androidx.compose.runtime.Composable

/**
 * **Android (Compose) feature wrapper.** Renders [content] only if [snapshot] has the feature
 * [featureId]; otherwise renders [fallback] (nothing by default). The coarse counterpart of
 * [CapabilityGate] (which checks a single capability) — use this to show/hide a whole feature/section.
 * The tab bar gates the same way in `App.kt` via `snapshot.hasFeature(...)`.
 *
 * ```
 * FeatureGate(snapshot, FeatureId.ORDERS) {
 *     OrdersSummaryCard(...)
 * }
 * ```
 */
@Composable
fun FeatureGate(
    snapshot: ExperienceSnapshot,
    featureId: FeatureId,
    fallback: @Composable () -> Unit = {},
    content: @Composable () -> Unit,
) {
    if (snapshot.hasFeature(featureId)) content() else fallback()
}

File: iosApp/iosApp/FeatureGate.swift
import SwiftUI
import Shared

/// **iOS (SwiftUI) feature wrapper** — the counterpart of the Android Compose `FeatureGate`, and the
/// coarse partner of `CapabilityGate`. Renders `content` only if `snapshot` has the feature `featureId`.
///
/// Shares the same Kotlin check: `ExperienceSnapshot.hasFeature` (exported from the `Shared` framework).
///
/// ```swift
/// FeatureGate(snapshot: snapshot, featureId: .orders) {
///     OrdersSummaryCard(...)
/// }
/// ```
struct FeatureGate<Content: View>: View {
    let snapshot: ExperienceSnapshot
    let featureId: FeatureId
    @ViewBuilder let content: () -> Content

    var body: some View {
        if snapshot.hasFeature(featureId: featureId) {
            content()
        }
    }
}
